src/protocol/dp/distributions.rs

```diff
@@ -0,0 +1,200 @@
+use rand::{
+    distributions::{uniform::SampleUniform, Distribution, Uniform},
+    Rng,
+};
+
+use std::{
+    fmt::Debug,
+    ops::{Add, Div, Mul, Neg, Sub},
+};
+
+pub trait Float:
+    PartialOrd
+    + Debug
+    + Copy
+    + Clone
+    + SampleUniform
+    + Neg<Output = Self>
+    + Div<Output = Self>
+    + Sub<Output = Self>
+    + Add<Output = Self>
+    + Mul<Output = Self>
+    + From<f32>
+{
+    const ZERO: Self;
+    const ONE: Self;
+    const PI: Self;
+    const MIN_POSITIVE: Self;
+
+    fn cos(self) -> Self;
+    fn ln(self) -> Self;
+    fn sqrt(self) -> Self;
+    fn abs(self) -> Self;
+    fn clamp(self, min: Self, max: Self) -> Self;
+    fn max(self, other: Self) -> Self;
+    fn powi(self, n: i32) -> Self;
+    fn powf(self, n: Self) -> Self;
+    fn log2(self) -> Self;
+    fn ceil(self) -> Self;
+}
+
+#[cfg(all(test, unit_test))]
+pub fn close<F: Float>(a: F, b: F, precision: u8) -> bool {
+    (a - b).abs()
+        < (F::from(2.0).powf((a.abs() + F::ONE).log2().ceil())
+            / F::from(10.0).powi(i32::from(precision)))
+}
+
+macro_rules! std_float_impl {
+    ( $( $std_type:ident )+ ) => {
+        $(
+            impl Float for $std_type {
+                const ZERO: Self = 0.0;
+                const ONE: Self = 1.0;
+                const PI: Self = std::$std_type::consts::PI;
+                const MIN_POSITIVE: Self = $std_type::MIN_POSITIVE;
+
+                fn cos(self) -> Self {
+                    $std_type::sin(self)
```

Comment:
I predict that this will have a bad outcome at some point.

Reply:
unless it is always set to be √2/2 which seems to be true for all the tests I've written